### PR TITLE
JIT fusion, with Memory size fusion policy

### PR DESCRIPTION
**Pull request by lessw2020**
This PR implements:
1 - JIT fusion - this is fusion where a local buffer is created for each fusion group, rather than using a global buffer.  
2 - a mb based policy for JIT instead of integer fusion (i.e. comms are fused once they exceed the MB size rather than a count of total comms to fuse)
3 - Removes overlap_communication optimization.  This was an interim optimization to disperse the comms but is now integrated into the full policies ala JIT, Ring, CAT.

### CODE REVIEW

**Comment by lessw2020** [comment]
For reference - there is a failing spmd test due to a missing rule in DTensor - not anything related to this PR:

File "/usr/local/lib/python3.9/site-packages/torch/distributed/_tensor/dispatch.py", line 244, in operator_dispatch
    target_schema, redistribute, output_sharding = propagate_input_sharding(
  File "/usr/local/lib/python3.9/site-packages/torch/distributed/_tensor/dispatch.py", line 183, in propagate_input_sharding
    raise NotImplementedError(
NotImplementedError: Operator aten.fill.Scalar does not have a DistributedTensor rule registered.

File: spmd/compiler/fusion.py [hunk 1]
```diff
@@ -970,3 +935,285 @@ def run_fuse_communication_cat(gm: fx.GraphModule, fusion_length: int) -> None:
...
+    """prepare views against completed buffer, these will replace gradient nodes for
+    graph output to avoid copy back overhead."""
+
+    assert scatter_list[-1].wait_node is not None
+    wait_node = scatter_list[-1].wait_node
+
+    # ensure user
+
+    wait_user = wait_node.args[0]  # type: ignore
+    wait_node.users[wait_user] = ""  # type: ignore
```
Comment (fegin) [inline spmd/compiler/fusion.py:1081]: Why assigning users an empty string instead of popping it out? Any why do we have to modify the `users`?

Comment (fegin) [reply spmd/compiler/fusion.py:1081]: Never mind. You mentioned that in the talk.

File: examples/spmd/run_spmd.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ def setup(rank: int, world_size: int, use_cuda: bool = True) -> None:
         _debug("--> init process group using nccl")
         dist.init_process_group("nccl", rank=rank, world_size=world_size)
         torch.cuda.set_device(rank)
-        print(f"--> device set for rank {rank}")
+        # print(f"--> device set for rank {rank}")
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply examples/spmd/run_spmd.py:35]: agree, removed.

File: examples/spmd/run_spmd.py [hunk 3]
```diff
@@ -119,7 +119,7 @@ def work_main(rank: int, world_size: int) -> None:
     _debug(f"mesh set to {mesh}\n")
 
     # control depth of ReplicaModel
-    layers = 4
+    layers = 5
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply examples/spmd/run_spmd.py:122]: ah, I was testing out both even fusions as well as the edge case of a fusion that didn't fit within the policy (assuming default int policy of 2). 
Let me revert it to 4.

File: spmd/compiler/fusion.py [hunk 4]
```diff
@@ -970,3 +935,285 @@ def run_fuse_communication_cat(gm: fx.GraphModule, fusion_length: int) -> None:
...
+
+
+def run_fuse_communication_jit(gm: fx.GraphModule, fusion_length: int) -> None:
+
+    """runs fusion by creating a Just in Time buffer to use for each fusion.
+    It then returns views to the buffer for the gradient outputs, avoiding the
+    need to copy back to the original tensor.
+    We can thus del the gradient tensors as fused, and by only creating buffers as
+    needed, minimize overhead memory pressure."""
+    MBFACTOR = float(1 << 20)
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply spmd/compiler/fusion.py:1139]: removed

File: spmd/compiler/fusion.py (hunk 1 above)
Comment (lessw2020) [reply spmd/compiler/fusion.py:1081]: cool, thanks.  
Just for any future viewing reference - certain graph operations such as copy do not register as having a user, so we have to manually assign it a user or it gets blown away during next recompile.  
The user 'dict' is treated as a set, so only keys count and value = "don't care" ,thus assigning it with "".

File: spmd/compiler/fusion.py [hunk 5]
```diff
@@ -970,3 +935,285 @@ def run_fuse_communication_cat(gm: fx.GraphModule, fusion_length: int) -> None:
...
+            torch.ops.aten.slice.Tensor,
+            (jit_buffer_node, 0, start, stop),
+        )
+
+        view_node = gm.graph.call_function(
+            torch.ops.aten.view.default,
+            (slice_node, shape),
+        )
+
+        """copy_out_node = gm.graph.call_function(
```
[the remark this answers is not in the document]
Comment (lessw2020) [reply spmd/compiler/fusion.py:1106]: yes good catch. removed.

**Review by fegin — APPROVED** [review]
LGTM

**Pull request merged**